### PR TITLE
CDAP-1609 CDAP-9251 CDAP-8948 Add get/list schedule handler, test cas…

### PR DESCRIPTION
**Pull request by rohitsinha54**
…es for add/update/get/delete schedule and bugfixes

### CDAP-1609
- Add get schedule api
- Add list schedules api
- Add unit tests for add/update/delete/get/list schedules

### CDAP-9251
- If the cron expression of schedule is invalid throw BadRequest for 40x response code rather than  IllegalArgumentException causing 50x response code

### CDAP-8948
- If the given programType is not Schedulable throw BadRequest for 40x response code rather than  IllegalArgumentException causing 50x response code

Build: http://builds.cask.co/browse/CDAP-RUT857-1

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java [hunk 1]
```diff
@@ -963,6 +1010,17 @@ public void deleteSchedule(ApplicationId applicationId, String scheduleName)
     return programRecords;
   }
 
+  private ProgramType getSchedulableProgramType(ScheduleSpecification scheduleSpec) throws BadRequestException {
+    try {
+      return ProgramType.valueOfSchedulableType(scheduleSpec.getProgram().getProgramType());
+    } catch (IllegalArgumentException e) {
+      // the ProgramType.valueOfSchedulableType throws IllegalArgumentException if the given program type is not in
+      // SchedulableProgramType just wrap is in BadRequest to send proper error code to user if this call is made from
```
Comment (Denton-L) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java]: wrap is -> wrap it

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java [hunk 2]
```diff
@@ -980,10 +980,14 @@ private static ApplicationSpecification updateFlowletInstancesInAppSpec(Applicat
...
-    ApplicationSpecification appSpec = getApplicationSpec(mds, id.getParent());
+    return getAppSpecOrFail(mds, id.getParent());
+  }
+
+  private ApplicationSpecification getAppSpecOrFail(AppMetadataStore mds, ApplicationId id) {
+    ApplicationSpecification appSpec = getApplicationSpec(mds, id);
     if (appSpec == null) {
       throw new NoSuchElementException("no such application @ namespace id: " + id.getNamespaceId() +
-                                           ", app id: " + id.getApplication());
+                                         ", app id: " + id.getApplication());
```
Comment (Denton-L) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:990]: nit: should be aligned with the quotes of the previous line

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/AppFabricTestBase.java [hunk 3]
```diff
@@ -863,6 +870,46 @@ protected int resumeSchedule(String namespace, String appName, String schedule)
     return GSON.fromJson(json, new TypeToken<List<ScheduleSpecification>>() { }.getType());
   }
 
+  protected HttpResponse addSchedule(String namespace, String appName, @Nullable String appVersion, String scheduleName,
+                             ScheduleSpecification scheduleSpec) throws Exception {
+    appVersion = appVersion == null ? ApplicationId.DEFAULT_VERSION : appVersion;
+    String path = String.format("apps/%s/versions/%s/schedules/%s", appName, appVersion, scheduleName);
+    return  doPut(getVersionedAPIPath(path, namespace), GSON.toJson(scheduleSpec));
```
Comment (Denton-L) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/AppFabricTestBase.java]: nit: shouldn't be two spaces after the return

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/AppFabricTestBase.java [hunk 4]
```diff
@@ -863,6 +870,46 @@ protected int resumeSchedule(String namespace, String appName, String schedule)
...
+    appVersion = appVersion == null ? ApplicationId.DEFAULT_VERSION : appVersion;
+    String path = String.format("apps/%s/versions/%s/schedules/%s", appName, appVersion, scheduleName);
+    return  doDelete(getVersionedAPIPath(path, namespace));
+  }
+
+  protected HttpResponse updateSchedule(String namespace, String appName, @Nullable String appVersion,
+                                        String scheduleName, ScheduleSpecification scheduleSpec) throws Exception {
+    appVersion = appVersion == null ? ApplicationId.DEFAULT_VERSION : appVersion;
+    String path = String.format("apps/%s/versions/%s/schedules/%s/update", appName, appVersion, scheduleName);
+    return  doPost(getVersionedAPIPath(path, namespace), GSON.toJson(scheduleSpec));
```
Comment (Denton-L) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/AppFabricTestBase.java]: nit: shouldn't be two spaces after the return

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/AppFabricTestBase.java [hunk 5]
```diff
@@ -863,6 +870,46 @@ protected int resumeSchedule(String namespace, String appName, String schedule)
...
+    appVersion = appVersion == null ? ApplicationId.DEFAULT_VERSION : appVersion;
+    String path = String.format("apps/%s/versions/%s/schedules/%s", appName, appVersion, scheduleName);
+    return  doPut(getVersionedAPIPath(path, namespace), GSON.toJson(scheduleSpec));
+  }
+
+  protected HttpResponse deleteSchedule(String namespace, String appName, @Nullable String appVersion,
+                                        String scheduleName) throws Exception {
+    appVersion = appVersion == null ? ApplicationId.DEFAULT_VERSION : appVersion;
+    String path = String.format("apps/%s/versions/%s/schedules/%s", appName, appVersion, scheduleName);
+    return  doDelete(getVersionedAPIPath(path, namespace));
```
Comment (Denton-L) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/AppFabricTestBase.java]: nit: shouldn't be two spaces after the return

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java [hunk 6]
```diff
@@ -927,6 +936,44 @@ public void deleteSchedule(ApplicationId applicationId, String scheduleName)
...
+    ScheduleSpecification scheduleSpec = appSpec.getSchedules().get(scheduleName);
+    if (scheduleSpec == null) {
+      throw new NotFoundException(new ScheduleId(applicationId.getNamespace(), applicationId.getApplication(),
+                                                 scheduleName));
+    }
+    return scheduleSpec;
+  }
+
+  /**
+   * Gets all the the schedules in a application
```
Comment (Denton-L) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java]: Gonna do JJ's job; this sentence should end with a period.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java [hunk 7]
```diff
@@ -927,6 +936,44 @@ public void deleteSchedule(ApplicationId applicationId, String scheduleName)
...
+   * @param applicationId the application id
+   * @return List of {@link ScheduleSpecification} of all the schedules in the application
```
Comment (Denton-L) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java]: L in List should be lowercase.

**Review by Denton-L — CHANGES_REQUESTED** [review]
Logic seems fine to me, just a few nitpicks about docs and code styling.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java (hunk 2 above)
Comment (rohitsinha54) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:990]: Its auto aligned. I think its making it inside ( Look at all the ones above.

Comment (Denton-L) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:990]: I see; my mistake!

**Review by Denton-L — APPROVED** [review]

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java [hunk 8]
```diff
@@ -927,6 +936,45 @@ public void deleteSchedule(ApplicationId applicationId, String scheduleName)
...
+    ScheduleSpecification scheduleSpec = appSpec.getSchedules().get(scheduleName);
+    if (scheduleSpec == null) {
+      throw new NotFoundException(new ScheduleId(applicationId.getNamespace(), applicationId.getApplication(),
+                                                 scheduleName));
+    }
+    return scheduleSpec;
+  }
+
+  /**
+   * Gets all the the schedules in a application.
```
Comment (Denton-L) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java]: Sorry, one more; change "the the" to "of the".

Comment (rohitsinha54) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java]: 👍

**Comment by rohitsinha54** [comment]
Random build failure here http://builds.cask.co/browse/CDAP-RUT857-3 
Running another time. Will merge as soon as it passes.

**Pull request merged**